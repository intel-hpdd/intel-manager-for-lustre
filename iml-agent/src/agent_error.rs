// Copyright (c) 2020 DDN. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

use iml_cmd::CmdError;
use iml_fs::ImlFsError;
use iml_wire_types::PluginName;
use std::{fmt, process::Output};
use thiserror::Error;
use tokio_util::codec::LinesCodecError;

pub type Result<T> = std::result::Result<T, ImlAgentError>;

#[derive(Debug)]
pub struct NoSessionError(pub PluginName);

impl fmt::Display for NoSessionError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "No session found for {:?}", self.0)
    }
}

impl std::error::Error for NoSessionError {
    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
        None
    }
}

#[derive(Debug)]
pub struct NoPluginError(pub PluginName);

impl fmt::Display for NoPluginError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "Plugin in registry not found for {:?}", self.0)
    }
}

impl std::error::Error for NoPluginError {
    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
        None
    }
}

#[derive(Debug)]
pub struct RequiredError(pub String);

impl fmt::Display for RequiredError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "{:?}", self.0)
    }
}

impl std::error::Error for RequiredError {
    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
        None
    }
}

#[derive(Debug)]
pub struct CibError(pub String);

impl fmt::Display for CibError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "{}", self.0)
    }
}

impl std::error::Error for CibError {
    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
        None
    }
}

#[derive(Debug, Error)]
pub enum ImlAgentError {
    #[error(transparent)]
    AddrParseError(#[from] std::net::AddrParseError),
    #[error(transparent)]
    IpNetworkError(#[from] ipnetwork::IpNetworkError),
    #[error(transparent)]
    CibError(#[from] CibError),
    #[error(transparent)]
    CmdError(#[from] CmdError),
    #[error(transparent)]
    FmtError(#[from] strfmt::FmtError),
    #[error(transparent)]
    FromUtf8Error(#[from] std::string::FromUtf8Error),
    #[error(transparent)]
    ImlFsError(#[from] ImlFsError),
    #[error(transparent)]
    InvalidHeaderValue(#[from] http::header::InvalidHeaderValue),
    #[error(transparent)]
    InvalidUri(#[from] http::uri::InvalidUri),
    #[error(transparent)]
    InvalidUriParts(#[from] http::uri::InvalidUriParts),
    #[error(transparent)]
    Io(#[from] std::io::Error),
    #[error(transparent)]
    LiblustreError(#[from] liblustreapi::error::LiblustreError),
    #[error(transparent)]
    LinesCodecError(#[from] LinesCodecError),
    #[error(transparent)]
    LustreCollectorError(#[from] lustre_collector::error::LustreCollectorError),
    #[error("Marker Not Found")]
    MarkerNotFound,
    #[error("Argument '{0}' Missing")]
    MissingArgument(String),
    #[error(transparent)]
    NoPluginError(#[from] NoPluginError),
    #[error(transparent)]
    NoSessionError(#[from] NoSessionError),
    #[error(transparent)]
    OneshotCanceled(#[from] futures::channel::oneshot::Canceled),
    #[error(transparent)]
    ParseBoolError(#[from] std::str::ParseBoolError),
    #[error(transparent)]
    ParseIntError(#[from] std::num::ParseIntError),
    #[error(transparent)]
    RequiredError(#[from] RequiredError),
    #[error(transparent)]
    Reqwest(#[from] reqwest::Error),
    #[error("Rx went away")]
    SendError,
    #[error(transparent)]
    Serde(#[from] serde_json::Error),
    #[error(transparent)]
    SystemdError(#[from] iml_systemd::SystemdError),
    #[error(transparent)]
    TokioJoinError(#[from] tokio::task::JoinError),
    #[error(transparent)]
    TokioTimerError(#[from] tokio::time::Error),
    #[error("Unexpected status code")]
    UnexpectedStatusError,
    #[error(transparent)]
    UrlParseError(#[from] url::ParseError),
    #[error(transparent)]
    Utf8Error(#[from] std::str::Utf8Error),
    #[error(transparent)]
    XmlError(#[from] elementtree::Error),
    #[error(transparent)]
    QuickXmlError(#[from] quick_xml::Error),
    #[error("{0}")]
    LdevEntriesError(String),
    #[error(transparent)]
    CombineEasyError(#[from] combine::stream::easy::Errors<char, String, usize>),
}

impl From<Output> for ImlAgentError {
    fn from(output: Output) -> Self {
        ImlAgentError::CmdError(output.into())
    }
}

impl From<futures::channel::mpsc::SendError> for ImlAgentError {
    fn from(_: futures::channel::mpsc::SendError) -> Self {
        ImlAgentError::SendError
    }
}

impl serde::Serialize for ImlAgentError {
    fn serialize<S>(&self, serializer: S) -> std::result::Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        serializer.serialize_str(&format!("{:?}", self))
    }
}
