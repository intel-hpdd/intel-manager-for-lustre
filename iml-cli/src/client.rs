// Copyright (c) 2020 DDN. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    iml_tracing::init();

    Ok(())
}
