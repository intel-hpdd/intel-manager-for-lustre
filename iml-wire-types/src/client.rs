// Copyright (c) 2020 DDN. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

//! Data structures for communicating with the agent regarding lustre client mounts.

#[cfg(feature = "cli")]
use structopt::StructOpt;

#[derive(serde::Deserialize, Debug)]
#[cfg_attr(feature = "cli", derive(StructOpt))]
/// Ask agent to mount a client
pub struct Mount {
    /// mountspec
    pub mountspec: String,
    /// mountpoint
    pub mountpoint: String,
    /// Persist mount across reboots
    #[cfg_attr(feature = "cli", structopt(short = "p", long = "persist"))]
    pub persist: bool,
}

#[derive(serde::Deserialize, Debug)]
#[cfg_attr(feature = "cli", derive(StructOpt))]
/// Ask agent to unmount a client and clear matching `/etc/fstab` entry
pub struct Unmount {
    /// mountspec
    pub mountspec: String,
    /// mountpoint
    pub mountpoint: String,
}
