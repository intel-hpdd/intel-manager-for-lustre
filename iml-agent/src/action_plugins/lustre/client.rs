// Copyright (c) 2020 DDN. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

use crate::agent_error::ImlAgentError;
use futures::TryFutureExt;
use iml_cmd::{CheckedCommandExt, Command};
use iml_fs::read_file_to_end;
use iml_wire_types::client::{Mount, Unmount};
use std::io;
use tokio::{
    fs::{self, File},
    io::AsyncWriteExt,
};

static FSTAB_EDIT: &str = "/etc/fstab.iml.edit";
static FSTAB: &str = "/etc/fstab";

/// Given a mountpoint, this fn will check if it is mounted via `findmnt`.
/// It can optionally check `fstab` instead of the kernel.
async fn is_mounted(mountpoint: &str, check_fstab: bool) -> Result<bool, ImlAgentError> {
    let mut cmd = Command::new("/usr/bin/findmnt");

    if check_fstab {
        cmd.arg("--fstab");
    }

    let output = cmd
        .args(vec![
            "--target",
            &mountpoint,
            "--output",
            "TARGET",
            "--noheadings",
        ])
        .output()
        .await?;

    match output.status.code() {
        Some(0) => {
            let output = std::str::from_utf8(&output.stdout)?;

            Ok(output.lines().any(|x| x.trim() == mountpoint))
        }
        Some(1) => Ok(false),
        Some(code) => Err(ImlAgentError::Io(io::Error::new(
            io::ErrorKind::Other,
            format!("is_mounted: {} exited with code: {}", mountpoint, code),
        ))),
        None => Err(ImlAgentError::Io(io::Error::new(
            io::ErrorKind::Other,
            format!("is_mounted: {} was cancelled", mountpoint),
        ))),
    }
}

/// This action will attempt to:
/// - Create the specified `Mount.mountpoint` path
/// - Mount the client
/// - Add a systemd mount to `/etc/fstab`
/// - Optionally set systemd mount to start at boot iff `Mount.persist` is `true`.
///
/// Ref: http://wiki.lustre.org/Mounting_a_Lustre_File_System_on_Client_Nodes
pub async fn mount(mount: Mount) -> Result<(), ImlAgentError> {
    if is_mounted(&mount.mountpoint, false).await? {
        tracing::info!(
            "Mount command called for {}, but it was already mounted",
            mount.mountpoint
        );

        return Ok(());
    }

    // This should return Ok(()) if the dir already exists.
    fs::create_dir_all(&mount.mountpoint).await?;

    let args = vec!["-t", "lustre", &mount.mountspec, &mount.mountpoint];
    Command::new("/bin/mount")
        .args(args)
        .checked_output()
        .await?;

    update_fstab_entry(&mount.mountspec, &mount.mountpoint, mount.persist).await?;

    Ok(())
}

pub async fn mount_many(xs: Vec<Mount>) -> Result<(), ImlAgentError> {
    for x in xs {
        mount(x).await?;
    }

    Ok(())
}

pub async fn unmount(unmount: Unmount) -> Result<(), ImlAgentError> {
    delete_fstab_entry(&unmount.mountpoint).await?;

    Command::new("/bin/umount")
        .arg(unmount.mountpoint)
        .checked_output()
        .err_into()
        .await
        .map(drop)
}

/// This action will attempt to:
/// - Create the specified `Mount.mountpoint` path
/// - Add a systemd mount to `/etc/fstab`
/// - Optionally set systemd mount to start at boot iff `Mount.persist` is `true`.
pub async fn add_fstab_entry(mount: Mount) -> Result<(), ImlAgentError> {
    // This should return Ok(()) if the dir already exists.
    fs::create_dir_all(&mount.mountpoint).await?;

    update_fstab_entry(&mount.mountspec, &mount.mountpoint, mount.persist).await?;

    Ok(())
}

/// This action will attempt to:
/// - Remove a mountpoint from `/etc/fstab`
pub async fn remove_fstab_entry(mount: Mount) -> Result<(), ImlAgentError> {
    delete_fstab_entry(&mount.mountpoint).await?;

    Ok(())
}

pub async fn unmount_many(xs: Vec<Unmount>) -> Result<(), ImlAgentError> {
    for x in xs {
        unmount(x).await?;
    }

    Ok(())
}

async fn get_fstab() -> Result<String, ImlAgentError> {
    let contents = read_file_to_end(FSTAB).await?;
    let contents = String::from_utf8(contents)?;

    Ok(contents)
}

fn filter_mountpoint(mountpoint: &str, contents: &str) -> String {
    contents
        .split('\n')
        .filter(|x| match x.split_whitespace().nth(1) {
            Some(x) => x != mountpoint,
            None => true,
        })
        .collect::<Vec<_>>()
        .join("\n")
}

async fn write_fstab(contents: &str) -> Result<(), ImlAgentError> {
    let mut file = File::create(FSTAB_EDIT).await?;
    file.write_all(contents.as_bytes()).await?;

    fs::rename(FSTAB_EDIT, FSTAB).await?;

    Ok(())
}

async fn update_fstab_entry(
    mountspec: &str,
    mountpoint: &str,
    persist: bool,
) -> Result<(), ImlAgentError> {
    let persist = if persist { "" } else { "noauto," };

    delete_fstab_entry(mountpoint).await?;

    let contents = get_fstab()
        .await?
        .split('\n')
        .chain(std::iter::once(
            format!(
                "{}\t{}\tlustre\tx-systemd.mount-timeout=20m,{}_netdev",
                mountspec, mountpoint, persist,
            )
            .as_str(),
        ))
        .collect::<Vec<_>>()
        .join("\n");

    write_fstab(&contents).await?;

    Ok(())
}

async fn delete_fstab_entry(mountpoint: &str) -> Result<(), ImlAgentError> {
    let contents = get_fstab().await?;
    let new_contents = filter_mountpoint(mountpoint, &contents);

    if contents != new_contents {
        write_fstab(&new_contents).await?;
    }

    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_filter() {
        let fstab = r#"# /etc/fstab
# blah

/dev/foo        /bar        something   defaults    1 1
1.2.3.4@tcp:/foobar              /mnt/lustre_clients/foobar          lustre      x-systemd.mount-timeout=20m,noauto,_netdev
1.2.3.4@tcp:/foobar1             /mnt/lustre_clients/foobar1         lustre      x-systemd.mount-timeout=20m,_netdev

1.2.3.4:/baz    /qux        nfs         defaults    0 0

"#;

        insta::assert_display_snapshot!(filter_mountpoint("/mnt/lustre_clients/foobar", fstab));
        insta::assert_display_snapshot!(filter_mountpoint("/mnt/lustre_clients/foobar1", fstab));

        let no_match = filter_mountpoint("qux", fstab);
        insta::assert_display_snapshot!(no_match);

        assert_eq!(fstab, no_match);
    }
}
