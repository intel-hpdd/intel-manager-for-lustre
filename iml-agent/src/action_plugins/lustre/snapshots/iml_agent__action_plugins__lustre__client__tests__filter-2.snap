---
source: iml-agent/src/action_plugins/lustre/client.rs
expression: "filter_mountpoint(\"/mnt/lustre_clients/foobar1\", fstab)"
---
# /etc/fstab
# blah

/dev/foo        /bar        something   defaults    1 1
1.2.3.4@tcp:/foobar              /mnt/lustre_clients/foobar          lustre      x-systemd.mount-timeout=20m,noauto,_netdev

1.2.3.4:/baz    /qux        nfs         defaults    0 0


